File: java/javalog/src/test/java/com/laixhe/javalog/CharStringTest.java
package com.laixhe.javalog;

import com.laixhe.javalog.demo.CharStringDemo;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.*;

/**
 * 字符与字符串测试（对应 Rust char_string.rs 的 #[cfg(test)] 练习题）。
 */
class CharStringTest {

    // 练习 1：trim + 替换逗号为空格 + 按空格切词
    @Test
    void exercise1_trim_replace_split() {
        String replaced = "  Hello,Rust!  ".trim().replace(",", " ");
        assertArrayEquals(new String[]{"Hello", "Rust!"}, replaced.split("\\s+"));
    }

    // 练习 2：判断是否为中文（CJK 统一汉字区间 U+4E00 ~ U+9FFF）
    @Test
    void exercise2_is_chinese_char() {
        char han = '中';
        assertTrue(han >= 0x4E00 && han <= 0x9FFF);
        assertFalse('A' >= 0x4E00 && 'A' <= 0x9FFF);
    }

    // 练习 3：format! 拼接 &str 和 String（不转移所有权）
    @Test
    void exercise3_format_borrows() {
        String slice = "Hello";
        String owned = " Rust";
        String combined = String.format("%s%s", slice, owned);
        assertEquals("Hello Rust", combined);
        // format 只借用，owned 仍可用（Java String 不可变，天然无所有权转移问题）
        assertEquals(" Rust", owned);
    }

    // 练习 4：length / codePointCount / 字节数 的区别
    @Test
    void exercise4_length_bytes_codepoints() {
        String text = "你好😀";
        assertEquals(4, text.length());                            // UTF-16 码元数
        assertEquals(3, text.codePointCount(0, text.length()));    // Unicode 码点数
        assertEquals(10, text.getBytes(StandardCharsets.UTF_8).length); // UTF-8 字节数
    }

    // 练习 5：字符串解析
    @Test
    void exercise5_string_parse() {
        assertEquals(666, Integer.parseInt("666"));
        assertEquals(88.88, Double.parseDouble("88.88"));
        assertThrows(NumberFormatException.class, () -> Integer.parseInt("not_a_number"));
    }

    // 运行完整 Demo
    @Test
    void runCharStringDemo() {
        CharStringDemo.stdChar();
        CharStringDemo.stdString();
        CharStringDemo.stringParse();
    }
}

File: php/phplog/tests/CharStringTest.php
<?php

namespace Laixhe\Phplog\Tests;

use Laixhe\Phplog\CharStringDemo;
use PHPUnit\Framework\TestCase;

/**
 * 字符与字符串测试（对应 Rust char_string.rs 的 #[cfg(test)] 练习题）。
 */
final class CharStringTest extends TestCase
{
    // 练习 1：trim + 替换逗号为空格 + 按空格切词
    public function testExercise1TrimReplaceSplit(): void
    {
        $replaced = str_replace(',', ' ', trim('  Hello,Rust!  '));
        $this->assertSame(['Hello', 'Rust!'], preg_split('/\s+/', $replaced));
    }

    // 练习 2：判断是否为中文（CJK 统一汉字区间 U+4E00 ~ U+9FFF）
    public function testExercise2IsChineseChar(): void
    {
        $codePoint = mb_ord('中');
        $this->assertTrue($codePoint >= 0x4E00 && $codePoint <= 0x9FFF);
        $this->assertFalse(mb_ord('A') >= 0x4E00 && mb_ord('A') <= 0x9FFF);
    }

    // 练习 3：sprintf 拼接（不改变原字符串）
    public function testExercise3FormatBorrows(): void
    {
        $slice = 'Hello';
        $owned = ' Rust';
        $combined = sprintf('%s%s', $slice, $owned);
        $this->assertSame('Hello Rust', $combined);
        $this->assertSame('Hello', $slice); // 原字符串不变
    }

    // 练习 4：length / 字符数 / 字节数的区别
    public function testExercise4LengthBytesChars(): void
    {
        $text = '你好😀';
        $this->assertSame(10, strlen($text));          // UTF-8 字节数
        $this->assertSame(3, mb_strlen($text, 'UTF-8')); // 字符数
    }

    // 练习 5：字符串解析
    public function testExercise5StringParse(): void
    {
        $this->assertSame(666, (int) '666');
        $this->assertSame(88.88, (float) '88.88');
        $this->assertFalse(filter_var('not_a_number', FILTER_VALIDATE_INT));
    }

    // 运行完整 Demo
    public function testRunCharStringDemo(): void
    {
        $this->expectNotToPerformAssertions();
        CharStringDemo::stdChar();
        CharStringDemo::stdString();
        CharStringDemo::stringParse();
    }
}
